src/ILCompiler.Compiler/src/IL/EvaluationStack.cs

```diff
@@ -0,0 +1,482 @@
+                case TypeFlags.Byte:
+                case TypeFlags.Boolean:
+                    return Value >= byte.MaxValue || Value < 0;
+                case TypeFlags.Int16:
+                    return Value >= short.MaxValue || Value <= short.MinValue;
+                case TypeFlags.UInt16:
+                case TypeFlags.Char:
+                    return Value >= ushort.MaxValue || Value < 0;
+                case TypeFlags.Int32:
+                    return Value >= int.MaxValue || Value <= int.MinValue;
+                case TypeFlags.UInt32:
+                    return Value >= uint.MaxValue || Value < 0;
+                case TypeFlags.Int64:
+                    return false;
+                case TypeFlags.UInt64:
+                    return Value < 0;
+                default:
+                    return true;
+            }
+        }
+    }
+
+    internal class FloatConstantEntry : ConstantEntry<double>
+    {
+        public FloatConstantEntry(double value, TypeDesc type = null) : base(StackValueKind.Float, value, type)
+        {
+        }
+
+        public override void Append(CppGenerationBuffer _builder)
+        {
+            // TODO: Handle infinity, NaN, etc.
+            if (Double.IsNaN(Value))
+            {
+                _builder.Append("nan(0)");
+                throw new NotImplementedException();
+            }
+            else if (Double.IsInfinity(Value) || Double.IsPositiveInfinity(Value) || Double.IsNegativeInfinity(Value))
+            {
+                // Negative Infinity can be encoded as 0xFFF0000000000000
+                // Positive Infinity can be encoded as 0x7FF0000000000000
+                throw new NotImplementedException();
+            }
+            else
+            {
+                _builder.Append(Value.ToStringInvariant());
+            }
+        }
+
+        public override StackEntry Duplicate()
+        {
+            return new FloatConstantEntry(Value, Type);
+        }
+    }
+
+    internal class VariableEntry : StackEntry
+    {
+        public Value Value;
+
+        public VariableEntry(StackValueKind kind, Value value, TypeDesc type = null) : base(kind, type)
+        {
+            Value = value;
+        }
+        public override void Append(CppGenerationBuffer _builder)
+        {
+            _builder.Append(Value.Name);
+        }
+
+        public override StackEntry Duplicate()
+        {
+            return new VariableEntry(Kind, Value, Type);
+        }
+
+        protected override void BuildRepresentation(StringBuilder s)
+        {
+            base.BuildRepresentation(s);
+            if (s.Length > 0)
+            {
+                s.Append(' ');
+            }
+            s.Append(Value.Name);
+
+            if (Value.Aux != null)
+            {
+                s.Append(' ');
+                s.Append(Value.Aux);
+            }
+        }
+    }
+
+    internal class InvalidEntry : StackEntry
```

Comment:
Sorry, I got confused about your initial comments which I thought applied to the `ToString()` implementation just above.

I chose InvalidEntry vs null to be more obvious. If you do a mistake and you see both null and InvalidEntry, you know that you have done something wrong. Otherwise you just don't know. If you are ok, I'll keep it, but have a singleton instance for this type to avoid useless allocation.